src/wasm/wasm-ir-builder.cpp

```diff
@@ -634,18 +758,42 @@ Result<> IRBuilder::visitEnd() {
     iff->ifFalse = *expr;
     iff->finalize(iff->type);
     push(maybeWrapForLabel(iff));
+  } else if (auto* tryy = scope.getTry()) {
+    tryy->body = *expr;
+    tryy->finalize(tryy->type);
+    push(maybeWrapForLabel(tryy));
+  } else if (Try * tryy;
+             (tryy = scope.getCatch()) || (tryy = scope.getCatchAll())) {
+    tryy->catchBodies.push_back(*expr);
+    tryy->finalize(tryy->type);
+    push(maybeWrapForLabel(tryy));
   } else {
     WASM_UNREACHABLE("unexpected scope kind");
   }
   return Ok{};
 }
 
-Result<Index> IRBuilder::getLabelIndex(Name label) {
+Result<Index> IRBuilder::getLabelIndex(Name label, bool inDelegate) {
```

Comment:
Can we possibly call `visitDelegate` after calling `visitEnd` for that `try` and avoid this complexity? Setting `delegateTarget` afterwards does not affect the result of `Try::finalize`.

Reply:
I think the problem there is that `visitDelegate` wouldn't know which `Try` to add a `delegateTarget` to because `visitEnd` will have popped the `Try` off the scope stack. I suppose we could try to go find the `Try` at the top of the `expressionStack` of the new top scope, but that also seems like a hack. I prefer this current hack because the weirdness is at least captured in function signatures. If we tried to find the `Try` on the scope stack, the leaky abstraction wouldn't be visible or documented in any interfaces.